text/1003-cli-naming.rst

```diff
@@ -19,110 +19,238 @@ specified.
+
+Are ``list-*`` commands useful?
+-------------------------------
+
+**Concern**: there are nine ``list-*`` sub-commands which is over 36%
+of the available sub-commands. They mirror REPL functionality for listing
+items in the current database, for example ``\la`` or ``\list-aliases``
+will list available aliases. Are all of them necessary?
+
+**Decision**: the CLI should keep listings of entities that it can
+directly manipulate with other sub-commands. This includes:
+
+* ``list-databases`` (through ``create-database``);
+
+* ``list-ports`` (through ``configure``); and
+
+* ``list-roles`` (through ``alter-role``).
+
+Other listings should be removed as they are redundant with REPL
+functionality. Instead, the following form should be enabled::
+
+  $ echo "\list-object-types" | edgedb
+
+This is consistent with behavior of, say, the ``sqlite3`` CLI.
+
+The REPL help for ``\\`` commands is overloaded
+-----------------------------------------------
+
+**Concern**: It is convenient to expose the EdgeDB CLI commands directly in
+REPL. This can greatly simplify administrative tasks when the DB
+administrator has full DB admin rights but yet can't access the server shell.
+Sadly, the presence of those commands under the ``\\`` prefix mixes them
+up with the REPL-specific escape commands like ``\last-error``.
+
+**Decision**: Expose CLI-specific commands in the REPL via the ``!`` prefix,
+similar to how IPython exposes shell commands::
+
+  >>> \list-databases
+  tutorial
+
+  >>> SELECT 1;
+  {1}
+
+  >>> !dump db tutorial
+  done
+
+This way the internal REPL help system is not overloaded with rarely
+needed help on CLI commands and would only show the list of convenient ``\``
+commands with a hint that ``!help`` or ``!h`` can be used to list
+all CLI options.
+
+Name-spacing ``\help`` and ``!help`` is good for the usability, because
+the latter set of commands is not going to be used as frequently as
+the former.
+
+Connection options are passed in an unnatural spot
+--------------------------------------------------
+
+**Concern**: Current ``edgedb`` command usage is defined as::
+
+  edgedb [FLAGS] [OPTIONS] [SUBCOMMAND]
+
+where ``[OPTIONS]`` is defined as::
+
+      --dsn <dsn>                   DSN for EdgeDB to connect to
+  -d, --database <database>         Database name to connect to
+  -H, --host <host>                 Host of the EdgeDB instance
+  -P, --port <port>                 Port to connect to EdgeDB
+  -I, --instance <instance>         Local instance name created with
+                                    `edgedb server init` to connect to
+                                    (overrides host and port)
+  -u, --user <user>                 User name of the EdgeDB user
+
+This sometimes creates unreadable commands through inconvenient syntax,
+for instance::
+
+  $ edgedb -d tutorial dump tutorial.edgedb
+
+**Decision**: Passing *connection* options should be allowed anywhere
```

Comment:
ACTIONABLE ITEM 8: allow passing connection options inside sub-commands

IIUC, this is currently impossible in `clap`. So we might need to drop it unless @1st1 decides it's very important. I find it pragmatic so I'd be +1 to have it if at all possible.

Reply:
I'd love this to be possible and I'm wondering how hard it would be to make it so.

Reply:
Both of you ignore what [I've already said](https://github.com/edgedb/rfcs/pull/14#discussion_r449470218):
1. We have conflicting options already (see the link).
2. `dump --all --database something` is weird
3. How would it work with `configure`, can you specify options on every level?
4. How would this work with `server`, can you specify options after `server`?
5. Do they work in slash commands `\dump --database xxx` ?
6. Someone have to inspect if there are any other conflicts than what I've listed. And only after we figure out what to do with all of them we can +1 on this item
7. Also what to do with `--tab-separated` and similar output options? Do we keep them on the top level? Do we propagate them to all the same commands? Do we propagate them only to commands that it makes sense to (e.g. `list-*` and `query` but not `dump/restore`

(Note: if answer on (3) or (4) or (5) is negative, which I would assume so at least for (4), clap's support of option propagation will not help us anyway, i.e. we have to do it manually).

Reply:
> Both of you ignore what I've already said:

Sorry, it's been a while since I last saw it. I remembered it as soon as I clicked on the link though.

> We have conflicting options already (see the link).

Maybe we rename those conflicting options? `create role` and `alter role` can use `--set-password` and `--set-password-from-stdin` which actually are more readable and don't conflict with connection options.

> `dump --all`

I think we opted for having a separate `dump-all` command, as it's output and functionality is quite different from the basic `dump`.

> How would it work with configure, can you specify options on every level?

It's two levels. `edgedb [connection config] <command> ...` or `edgedb <command> [connection config] ...`. No need to use it for further nested subcommands.

The reason for that is that at least I view `edgedb dump` as an atomic command on its own, so it feels _really weird_ to me to say `edgedb -d db dump` and not `edgedb dump -d db`.

> Do they work in slash commands \dump --database xxx ?

Why not? Some slash commands already have a bunch of `--options`.

> Someone have to inspect if there are any other conflicts than what I've listed. And only after we figure out what to do with all of them we can +1 on this item

@ambv can you take a quick look?

> Also what to do with --tab-separated and similar output options? Do we keep them on the top level? Do we propagate them to all the same commands? Do we propagate them only to commands that it makes sense to (e.g. list-* and query but not dump/restore

I only care about connection options specifically (database name, host, etc).  IMO things like `--tab-separated` should follow the command, so it's going to be `edgedb <subcommand> --tab-separated`, if it supports it.

Reply:
> Why not? Some slash commands already have a bunch of --options.

A couple of points:
1. I wanted to avoid:
    ```
    some_database> \list-object-types -d other_database
    ```
2. And also avoid listing a bunch of options in `\list-object-types --help` (because there are two useful options and a much more connection options)
3. Options like `--password` are weird here.
4. Using `-d database` where `--password` were specified on command-line is also weird (users might expect password to be reused? Or that password is asked again?)

So generally because this is how repl works: all commands are executed on the same connection.

> I think we opted for having a separate dump-all command, as it's output and functionality is quite different from the basic dump.

I don't see it mentioned anywhere. @ambv ?

But it also doesn't solve the issue:
```
edgedb dump -u my_user -d my_database file.dump
```
Looks perfectly fine. But:
```
edgedb dump-all -u my_user -d my_database file.dump
```
Looks weird, but *may be useful* if there the `edgedb` database was dropped.

> I only care about connection options specifically (database name, host, etc). IMO things like --tab-separated should follow the command, so it's going to be edgedb <subcommand> --tab-separated, if it supports it.

I'm asking because they are also top-level options, that in turn mostly because they are options for non-interactive scripting (and that mode invoked without any subcommand). And they can easily be applicable to almost all the same commands as connection options (sans dump/restore), i.e. `edgedb --json migration-log`, mutation commands can also output status in JSON like `edgedb --json drop-role` (although the use case for latter is not clear).

So the general approach is that we don't propagate all top-level options. We just hand-picking some of them to some commands?

Reply:
>> I think we opted for having a separate dump-all command, as it's output and functionality is quite different from the basic dump.
>
> I don't see it mentioned anywhere. @ambv ?

I think we discussed it on one of our calls. Perhaps @elprans remembers? But I'm strongly in favor of making dumpall a separate CLI command.

Reply:
> So the general approach is that we don't propagate all top-level options. We just hand-picking some of them to some commands?

Yeah, that's what I'm thinking we should do. Generally, we'll have some options that are common and can be applied in subcommands too, and some will be REPL-specific (and it should be an error to try to use a REPL-specific option on a subcommand).

Reply:
> > Looks weird, but may be useful if there the edgedb database was dropped.
> 
> I'm not sure I follow that example. You still need to connect to the DB to dump all DBs that you have access to. Excluding a database from dump-all can be a valid use case, but I'd go for something like `--exclude` instead.

Your confusion is exactly the confusion I'm taking about. `-d` isn't about including or excluding databases from `dump-all` it's connection option at the wrong place (it's only useful if you have dropped `edgedb` because otherwise, you there is no reason to specify `-d`, default would work).

Reply:
I'll be merging this PR with this section marked as "Under Consideration" as it looks like it needs more discussion.

Two comments:
1. Now I remember talks about "dump-all" and agree it should be a separate command. We can bikeshed the name though because IIUC "dump-all" is more like "dump-everything", including role passwords and port configuration, right? I would put it under `edgedb server dump` as this is where instance-level commands belong.
2. Yes, we have command-line option conflicts and that's ironically another reason why I'm +1 to this change. Why? Because those are confusing, especially when you're splitting a long command invocation in a Bash script or Dockerfile into multiple backslash-delimited lines. It's relatively easy to misread an option then or to edit a command and put an option in the wrong place. Removing all those conflicts would be ideal in my opinion.

Reply:
> Now I remember talks about "dump-all" and agree it should be a separate command. We can bikeshed the name though because IIUC "dump-all" is more like "dump-everything", including role passwords and port configuration, right? I would put it under edgedb server dump as this is where instance-level commands belong.

I'm -1 on making it a `server` command. It works on any instance, not just local instances. Even more it's currently the only way for upgrading a remote instance.

We can do `dump-server` if `dump-all` is bad, but it may be confusing.

But dictionary says that "all" may be used as a synonym for "everything". So I think it short and clear enough.

Reply:
> Yes, we have command-line option conflicts and that's ironically another reason why I'm +1 to this change. Why? Because those are confusing, especially when you're splitting a long command invocation in a Bash script or Dockerfile into multiple backslash-delimited lines. 

+1, good point.

Reply:
> We can do dump-server if dump-all is bad, but it may be confusing.

It's `pg_dumpall` in Postgres, and IMO it would work just as good for us.